src/OpenSage.LowLevel/Graphics2D/DrawingContext.D2D.cs

```diff
@@ -84,16 +99,76 @@ private void PlatformFillRectangle(in RawRectangleF rect, in ColorRgbaF fillColo
+            {
+                var geometry = new PathGeometry(_graphicsDevice.DeviceContext.Factory);
+                var sink = geometry.Open();
+                sink.BeginFigure(ToRawVector2(tri.X1, tri.Y1), FigureBegin.Filled);
+                sink.SetFillMode(SharpDX.Direct2D1.FillMode.Winding);
+                sink.AddLine(ToRawVector2(tri.X2, tri.Y2));
+                sink.AddLine(ToRawVector2(tri.X3, tri.Y3));
+                sink.EndFigure(FigureEnd.Closed);
+                sink.Close();
+                _graphicsDevice.DeviceContext.FillGeometry(geometry, brush);
+            }
+        }
+
+        private void PlatformFillTriangle(in RawTriangleF triangle, Texture texture, in RawMatrix3x2 transform)
+        {
+            if (texture.MipMapCount > 1)
+                throw new InvalidDataException("Texture is not allowed to have Mipmaps for Direct2D!");
+
+            using (var dxgiSurface = texture.DeviceResource.QueryInterface<Surface>())
+            using (var bitmap = new Bitmap1(_graphicsDevice.DeviceContext, dxgiSurface))
+            using (var brush = CreateImageBrush(bitmap, transform))
+            {
+                var geometry = new PathGeometry(_graphicsDevice.DeviceContext.Factory);
+                var sink = geometry.Open();
+                sink.SetFillMode(SharpDX.Direct2D1.FillMode.Winding);
+                sink.BeginFigure(ToRawVector2(triangle.X1, triangle.Y1), FigureBegin.Filled);
+                sink.AddLine(ToRawVector2(triangle.X2, triangle.Y2));
+                sink.AddLine(ToRawVector2(triangle.X3, triangle.Y3));
+                sink.EndFigure(FigureEnd.Closed);
+                sink.Close();
+                _graphicsDevice.DeviceContext.FillGeometry(geometry, brush);
+            }
+        }
+
         private SolidColorBrush CreateBrush(in ColorRgbaF color)
         {
             return new SolidColorBrush(_graphicsDevice.DeviceContext, color.ToRawColor4());
         }
 
+        private BitmapBrush1 CreateImageBrush(in Bitmap1 image, in RawMatrix3x2 transform)
+        {
+            var brushProp = new BrushProperties()
+            {
+                Transform = ToRawMatrix3x2(transform),
+                Opacity = 1.0f
+            };
+
+            var brush = new BitmapBrush1(_graphicsDevice.DeviceContext, image, brushProp);
+
+            return brush;
+        }
+
         private static SharpDX.Mathematics.Interop.RawRectangleF ToRawRectangleF(in RawRectangleF value)
         {
             return new SharpDX.Mathematics.Interop.RawRectangleF(value.X, value.Y, value.X + value.Width, value.Y + value.Height);
         }
 
+        private static SharpDX.Mathematics.Interop.RawMatrix3x2 ToRawMatrix3x2(in RawMatrix3x2 m)
+        {
+            return new SharpDX.Mathematics.Interop.RawMatrix3x2(m.M11, m.M12, m.M21, m.M22, m.M31, m.M32);
+        }
+
+        private static SharpDX.Mathematics.Interop.RawVector2 ToRawVector2(in float x, in float y)
```

Comment:
No need for `in` on primitive types - on a 64-bit platform, it will actually end up using *more* memory (64-bit pointer vs 32-bit `float`).